api/src/main/java/org/openmrs/obs/handler/ImageHandler.java

```diff
@@ -131,60 +126,44 @@ public String[] getSupportedViews() {
 	 */
 	@Override
 	public Obs saveObs(Obs obs) throws APIException {
-		// Get the buffered image from the ComplexData.
-		BufferedImage img = null;
-		
-		Object data = obs.getComplexData().getData();
-		if (data instanceof BufferedImage) {
-			img = (BufferedImage) obs.getComplexData().getData();
-		} else if (data instanceof byte[]) {
-			  ByteArrayInputStream bis = new ByteArrayInputStream((byte[]) data);
-			    try {
-			      img = ImageIO.read(bis);
-			    }
-			    catch (IOException e) {
-			      throw new APIException("Obs.error.unable.convert.complex.data", new Object[] { "input stream" }, e);
-			    }
+		try {
+			String[] splitTitle = obs.getComplexData().getTitle().split("\\|");
+			String filename = splitTitle[0];
+			if (splitTitle.length > 1) {
+				filename = splitTitle[1];
 			}
-		 else if (data instanceof InputStream) {
-			try {
-				img = ImageIO.read((InputStream) data);
+			String extension = FilenameUtils.getExtension(filename);
+			String assignedKey = storageService.saveData((out) -> {
+				Object data = obs.getComplexData().getData();
+				
+				InputStream in = null;
+				if (data instanceof byte[]) {
+					in = new ByteArrayInputStream((byte[]) data);
+				} else if (data instanceof InputStream) {
+					in = (InputStream) data;
+				}
+				
+				BufferedImage img = null;
+				if (in != null) {
+					img = ImageIO.read(in);
+				} else if (data instanceof BufferedImage) {
+					img = (BufferedImage) data;
+				}
+				
 				if (img == null) {
-					throw new IllegalArgumentException("Invalid image file");
+					throw new APIException("Obs.error.cannot.save.complex", new Object[] { obs.getObsId() });
 				}
-			}
-			catch (IOException e) {
-				throw new APIException("Obs.error.unable.convert.complex.data", new Object[] { "input stream" }, e);
-			}
-		}
-		
-		if (img == null) {
-			throw new APIException("Obs.error.cannot.save.complex", new Object[] { obs.getObsId() });
-		}
-		
-		File outfile = null;
-		try {
-			outfile = getOutputFileToWrite(obs);
-			
-			String extension = getExtension(obs.getComplexData().getTitle());
-			
-			// TODO: Check this extension against the registered extensions for validity
-			
-			// Write the file to the file system.
-			ImageIO.write(img, extension, outfile);
-			
+				ImageIO.write(img, extension, out);
+				out.flush();
+			}, ObjectMetadata.builder().setFilename(filename).build(), getObsDir());
+
 			// Set the Title and URI for the valueComplex
-			obs.setValueComplex(extension + " image |" + outfile.getName());
-			
+			obs.setValueComplex(extension + " image |" + assignedKey);
+
 			// Remove the ComlexData from the Obs
 			obs.setComplexData(null);
-			
-		}
-		catch (IOException ioe) {
-			if (outfile != null && outfile.length() == 0) {
-				outfile.delete(); // OpenJDK 7 & 8 may leave a 0-byte file when ImageIO.write(..) fails.
```

Comment:
Should i assume that the above is no longer a possibility?

Reply:
I'll add some logic and test to make sure that file is deleted if exception occurs.